### PR TITLE
fix security exception when validating plugin package

### PR DESCRIPTION
**Pull request by 3nids**

### CODE REVIEW

File: qgis-app/plugins/validator.py
Comment (3nids) [reply qgis-app/plugins/validator.py]: this should be fixed

**Comment by sumandari** [comment]
thanks for fixing @3nids, I tried it in my local machine but it's passing zipfile contains forbidden folder. 

when I refactored it, i found that os.path.basename in the code will return the filename
```
>>> os.path.basename("dummyzip/icons/go_to_sample.svg")
'go_to_sample.svg'

>>> os.path.basename("dummyzip/__MACOSX/")
''
```

**Comment by 3nids** [comment]
should be better now! sorry for the noise.

**Comment by sumandari** [comment]
@3nids `not_allowed_dir` will raise error since it has not defined. it was from previous code.
130_validation

File: qgis-app/plugins/validator.py
Comment (m-kuhn) [inline qgis-app/plugins/validator.py:87]:
```suggestion
                                        % (forbidden_dir,)) )
```

File: qgis-app/plugins/validator.py
Comment (agiudiceandrea) [inline qgis-app/plugins/validator.py:84]:
```suggestion
            if forbidden_dir in zname.split('/'):
```

**Review by agiudiceandrea — CHANGES_REQUESTED** [review]
In ZIP file archives the separator is OS independent and is always a slash '/' (see 4.4.17 in [ZIP app note](https://pkware.cachefly.net/webdocs/casestudies/APPNOTE.TXT) and [zipfile source code](https://github.com/python/cpython/blob/170dec3598ec945c3f32ce053737f378c2ed1edc/Lib/zipfile.py#L351-L355)).
So using os.sep will not work on Windows, in case someone wants to use this script to check his zip file before submitting it.

**Comment by agiudiceandrea** [comment]
Anyway I think a better approach could be to check that the zip file must contain one and only one folder, named after the plugin name, at the root level and no other file or folder at the root level (see [16.5.3.4. Plugin structure](https://docs.qgis.org/testing/en/docs/pyqgis_developer_cookbook/plugins/releasing.html#plugin-structure)).

This will prevent https://github.com/qgis/QGIS-Django/issues/62 without explicitly check for the '__MACOSX' (or others) folder (which creates the issue only when is at the root level) and it will also prevent other various issues and security issues:
```python
if os.path.commonprefix(zip.namelist()).find('/') == -1: # if commonprefix is empty or it's not a folder
    raise ValidationError( _("For security reasons, the zip file must contain only one directory, and no other files or directories, in the top level.") )
```
The code later [checks](https://github.com/sumandari/QGIS-Django/blob/master/qgis-app/plugins/validator.py#L161-L162) that the top level folder name must start with an ASCII letter, so it also prevents that the unique top level folder name could be __MACOSX or .git or __pycache__.

Eventually it could be useful to avoid .git or __pycache__ directories inside the unique top level directory, but it seems to me this is not a security issue: it could only prevent a waste of space.

However, I think it could be better to put these kind of checks after the [bad zip file checking](https://github.com/sumandari/QGIS-Django/blob/3be6043a2d2c552762c65485a17d57a66196b517/qgis-app/plugins/validator.py#L88-L95).

**Comment by 3nids** [comment]
@agiudiceandrea that makes sense. Are you going on with this? cheers.

**Comment by Joonalai** [comment]
What is the status of this PR?

**Comment by 3nids** [comment]
This is in a mergeable state.
Can we merge it if there is no plan to follow the other proposed approach?

**Comment by agiudiceandrea** [comment]
I think my proposed alternative code should be tested against the plugins zip files already present on the server to evaluate any false positives or false negatives before eventually use it in production. So better to merge this PR for now.

Anyway, I also think that the zip files on the server (at least only for the latest version of QGIS 3 plugins) should be checked after the merge like it was done last year: see http://osgeo-org.1560.x6.nabble.com/QGIS-Developer-Red-MACOSX-in-plugin-layer-list-td5397873.html

**Comment by Gustry** [comment]
Can we merge this PR ?

It seems some people are still having issue for uploading a plugin ? CC #147

**Comment by sumandari** [comment]
Hi @agiudiceandrea I gave it a try on my local machine against plugins zip files contains .gitignore and it works as expected, and also for another validation (will reject the zipfile if it contains .git, `__pycache__`, __MACOSX folder).

> I think my proposed alternative code should be tested against the plugins zip files already present on the server to evaluate any false positives or false negatives before eventually use it in production. So better to merge this PR for now.
> 
> Anyway, I also think it could be useful that the zip files on the server (at least only for the latest version of QGIS 3 plugins) be checked after the merge like it was done last year: see http://osgeo-org.1560.x6.nabble.com/QGIS-Developer-Red-MACOSX-in-plugin-layer-list-td5397873.html

**Comment by agiudiceandrea** [comment]
> Hi @agiudiceandrea I gave it a try on my local machine against plugins zip files contains .gitignore and it works as expected

Yes, indeed, I didn't doubt that.

I just want to point out that it could be useful to also search for not compliant plugins zip files already on the server and try to contact the authors in order to ask them to remove the forbidden folders in the zip files of their plugins like was done last year: see http://osgeo-org.1560.x6.nabble.com/QGIS-Developer-Red-MACOSX-in-plugin-layer-list-td5397873.html

**Comment by sumandari** [comment]
Well noted @agiudiceandrea. Thanks for your advice, I will check the plugins zip files already on server in another ticket. I will merge this PR and will let you know once it deployed.

**Comment by sumandari** [comment]
Please be informed that the PR has been deployed. Thanks all!

**Pull request merged**
